In [1]:
from sklearn.metrics import mean_squared_log_error
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import pandas as pd

In [2]:
df = pd.read_csv("training_dataset3.csv")

In [3]:
X = df[['Num_escuelas',
 'Num_hospitales',
 'Num_iglesias',
 'Num_zonas_tipicas',
  "Hora_llegada_1",                                          
  "Hora_llegada_2",
  "Hora_llegada_3",                                          
  "Hora_llegada_4",
  "Hora_llegada_5",                                          
  "Hora_llegada_6",
  "Hora_llegada_7",                                          
  "Hora_llegada_8",
  "Hora_llegada_9",                                          
  "Hora_llegada_10",
  "Hora_llegada_11",                                          
  "Hora_llegada_12",
  "Hora_llegada_13",                                          
  "Hora_llegada_14",
  "Hora_llegada_15",                                          
  "Hora_llegada_16",                                         
  "Hora_llegada_17",                                          
  "Hora_llegada_18",
  "Hora_llegada_19",                                          
  "Hora_llegada_20",
  "Hora_llegada_21",                                          
  "Hora_llegada_22",
  "Hora_llegada_23", ]]

y = df["Num_viajes"]

In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.15, random_state=42)

In [5]:
rforest_model = RandomForestRegressor()
rforest_model.fit(X, y)


RandomForestRegressor()

In [6]:
from sklearn.metrics import mean_absolute_error as mae
from sklearn.metrics import mean_squared_log_error as msle
import numpy as np 
yfit_rforest = pd.DataFrame(rforest_model.predict(X_train), index = X_train.index, columns = ["Num_viajes"]).clip(0.)

ypred_rforest = pd.DataFrame(rforest_model.predict(X_test), index = X_test.index, columns = ["Num_viajes"]).clip(0.)

print('Random forest MSLE :', np.sqrt(msle(y_test, ypred_rforest)))
print('Random forest MAE :', mae(y_test, ypred_rforest))

Random forest MSLE : 0.213621674976834
Random forest MAE : 78.9569060846561


In [7]:
dataframe = pd.DataFrame()

In [8]:
dataframe['Num_viajes'] = y_test

In [9]:
dataframe['Num_viajes_p'] = ypred_rforest

In [10]:
dataframe['dif'] = abs(dataframe['Num_viajes'] - dataframe['Num_viajes_p'])

# Evaluación del modelo

A continuación, calcularemos el porcentaje promedio de desviación entre la predicción y el valor real

In [11]:
import math

In [12]:
dataframe['Num_viajes_p'] =  dataframe['Num_viajes_p'].apply(lambda x: math.ceil(x))

In [13]:
dataframe["Dif_perc"] =  dataframe["dif"] / dataframe["Num_viajes"] *100

In [14]:
dataframe["Dif_perc"].mean()

15.046401407989643

Vemos que tenemos una desviación media de 15%, es decir, en promedio esta será la diferencia entre los viajes que se predicen y los que ocurren en realidad